publish.py

```diff
@@ -1,44 +1,98 @@
+import sys
+import time
+from typing import Optional
 
-MAPPINGS_DIR = 'mappings'
-GRADLEW = 'gradlew' if os.name == 'nt' else './gradlew'
 
-# some jank to hide versions that are giving problems
-UNAVAILABLE_VERSIONS = []
+class GradleProcess:
+    def __init__(self, version: str):
+        self._PUBLISH_TEMP_DIR: str = "publish_temp"
+        self._version: str = version
+        self._gradle_command: str = "gradlew" if os.name == "nt" else "./gradlew"
+        self._process: Optional[subprocess.Popen] = None
+        self._times_run: int = 0
+        self._TO_COPY: list[str] = ["gradle", "mappings", "build.gradle", "gradle.properties", "gradlew", "gradlew.bat", "settings.gradle"]
 
-def main():
-    versions = find_minecraft_versions()
-    
-    while versions:
-        failed = []
-        
-        for version in versions:
-            os.environ['MC_VERSION'] = version
+    def __enter__(self):
+        if not os.path.exists(self._PUBLISH_TEMP_DIR):
+            os.mkdir(self._PUBLISH_TEMP_DIR)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if os.path.exists(self._PUBLISH_TEMP_DIR):
+            shutil.rmtree(self._PUBLISH_TEMP_DIR)
+
+    def start(self) -> None:
+        while self._should_retry():
+            self._times_run += 1
+            print(f"Version {self._version}")
+            self._delete_dir()
+            os.mkdir(os.path.join(self._PUBLISH_TEMP_DIR, self._version))
+            self._get_ready()
+            self._run_gradle_command()
+            self._process_listener()
+            while self._process.poll() is None:
+                time.sleep(0.5)
+            self._delete_dir()
+            if self.get_exit_code() == 0:
+                return
+
+    def _should_retry(self) -> bool:
+        return self._times_run <= 2
+
+    def _delete_dir(self) -> None:
+        if os.path.isdir(os.path.join(self._PUBLISH_TEMP_DIR, self._version)):
+            shutil.rmtree(os.path.join(self._PUBLISH_TEMP_DIR, self._version))
+
+    def _get_ready(self) -> None:
+        for element in self._TO_COPY:
+            if os.path.isdir(element):
+                shutil.copytree(element, os.path.join(self._PUBLISH_TEMP_DIR, self._version, element))
+            else:
+                shutil.copy(element, os.path.join(self._PUBLISH_TEMP_DIR, self._version, element))
+
+    def get_exit_code(self) -> int:
+        return self._process.returncode
+
+    def get_version(self) -> str:
+        return self._version
+
+    def _run_gradle_command(self) -> None:
+        env = os.environ.copy()
+        env["MC_VERSION"] = self._version
+        # shell=True, check=True, env=env
+        self._process = subprocess.Popen(
+            f"{self._gradle_command} publish --stacktrace",
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            shell=True,
+            env=env,
+            cwd=os.path.join(self._PUBLISH_TEMP_DIR, self._version)
```

Comment:
What is the thought behind changing the working directory? I don't think it causes any issues, since Gradle will make sure any file access in the build script is relative to the project directory, but then this change has no effect at all?

Reply:
gradle doesn't allow doing multiple builds at the same time since it only uses one, so if we try to multithread it, it will break because "the same resource is being accessed from multiple processes", so i decided to just move the files gradle needs to survive to another directory named like the version and do the build there
It doesn't affect at all because there is only one process as you can see here https://github.com/OrnitheMC/feather-mappings/pull/134/files#diff-3252d091f3490b6e1103b1e47ed6e4d909cb8d2bb279bc54730383e0860ea566R184 but to keep the mutithread support for the future, this doesnt change anything, its more relevant for build than to publish, i can remove it from publish.py iyw

Reply:
> gradle doesn't allow doing multiple builds at the same time since it only uses one, so if we try to multithread it, it will break because "the same resource is being accessed from multiple processes", so i decided to just move the files gradle needs to survive to another directory named like the version and do the build there

ahhh I see, so it's needed to make the build_all work. 

> this doesnt change anything, its more relevant for build than to publish, i can remove it from publish.py iyw

at that point might as well just remove the publish script and publish through the feather script tbh

Reply:
is not exactly the same, i'd need to change the feather.py behaviour for the publish option to add the retry logic

or we don't want that anymore and just use the job action retry manually?

Reply:
the manual retry is probably good enough tbh

Reply:
okay~